src/jacobi/_jacobi.py

```diff
@@ -1 +1,261 @@
+            raise ValueError("step[1] must be between 0 and 1")
+    if method is not None and method not in (-1, 0, 1):
+        raise ValueError("method must be -1, 0, 1")
+
+    x = np.asarray(x, dtype=float)
+    if mask is not None:
+        mask = np.asarray(mask)
+        if mask.dtype != bool:
+            raise ValueError("mask must be a boolean array")
+        if mask.shape != x.shape:
+            raise ValueError("mask shape must match x shape")
+
+    if diagnostic is not None:
+        diagnostic["method"] = np.zeros(x.size, dtype=np.int8)
+        diagnostic["iteration"] = np.zeros(x.size, dtype=np.uint8)
+        diagnostic["residual"] = [[] for _ in range(x.size)]
+
+    f0 = None
+    jac = None
+    err = None
+    it = np.nditer(x, flags=["c_index", "multi_index"])
+    while not it.finished:
+        k = it.index
+        kx = it.multi_index if it.has_multi_index else ...
+        if mask is not None and not mask[kx]:
+            it.iternext()
+            continue
+        xk = it[0]
+        # if step is None, use optimal step sizes for central derivatives
+        h = _steps(xk, step or (0.25, 0.5), maxiter)
+        # if method is None, auto-detect for each x[k]
+        md, f0, r = _first(method, f0, fn, x, kx, h[0], args)
+        # f0 is not guaranteed to be set here and can be still None
+
+        if md != 0 and step is None:
+            # need different step sizes for forward derivatives to avoid overlap
+            h = _steps(xk, (0.25, 0.125), maxiter)
+
+        r = np.asarray(r, dtype=float)
+        re = np.full_like(r, np.inf)
+        todo = np.ones_like(r, dtype=bool)
+        fd = [np.reshape(r.copy(), -1)]
+
+        if jac is None:  # first iteration
+            jac = np.zeros(r.shape + x.shape, dtype=r.dtype)
+            err = np.zeros(r.shape + x.shape, dtype=r.dtype)
+            if diagnostic is not None:
+                diagnostic["call"] = np.zeros((r.size, x.size), dtype=np.uint8)
+
+        if diagnostic is not None:
+            diagnostic["method"][k] = md
+            diagnostic["call"][:, k] = 2 if md == 0 else 3
+
+        for i in range(1, len(h)):
+            fdi = _derive(md, f0, fn, x, kx, h[i], args)
+            fd.append(np.reshape(fdi, -1) if i == 1 else fdi[todo])
+            if diagnostic is not None:
+                diagnostic["call"][todo, k] += 2
+                diagnostic["iteration"][k] += 1
+
+            # polynomial fit with one extra degree of freedom;
+            # use latest maxgrad + 1 data points
+            grad = min(i - 1, maxgrad)
+            start = i - (grad + 1)
+            stop = i + 1
+            q, c = np.polyfit(
+                h[start:stop] ** 2, fd[start:], grad, rcond=None, cov=True
+            )
+            ri = q[-1]
+            # pulls have roughly unit variance, however,
+            # the pull distribution is not gaussian and looks
+            # more like student's t
+            rei = c[-1, -1] ** 0.5
```

Comment:
It seems like you're using this idea:

![image](https://github.com/mdhaber/jacobi/assets/6570539/c2295200-ac2e-44ca-a2e6-f964e0aa3de9)

so I was expecting the use of Student's t distribution, but I don't see it. Is the idea that you're just reporting the standard error instead of the size of a confidence interval?

Reply:
I am not following the original recipe, because the Students-t distribution is not exact for this case, since it assumes Gaussian fluctuations for the residuals. The residuals in this case come from both round-off error and truncation error, which are not Gaussian. Their probability distribution is not well-defined.

I chose am empirical approach and tried out several examples. I could not find an estimate for the error which is generally satisfactory, so I chose the simplest possible calculation (which is also the fastest) and just return the standard error reported by the fit. The error is an order of magnitude estimate than a correct standard error.

Reply:
Ok. Yeah I thought the multivariate t idea may have been practical but I prefer not to abuse statistics, too. Just checking.